### PR TITLE
Wenjie & Kat - Octos - Scrabble

### PR DESCRIPTION
**Pull request by kseastman**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | We can work independantly, and when we use git pull we can see the changes and decide which version handles the need better.
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | Because we never instantiate "Scoring" into a definite object, so it needs to be a class method, since it is functioning as a collection of methods, not as a discrete object.
Describe an example of a test you wrote for a _nominal case_. | In the Scoring class, we wrote a test to return the highest word if there were two words, this is nominal because it is making sure that code responds correctly to anticipated correct input.
Describe an example of a test you wrote for an _edge case_. | Using the Scoring class as an example, in the highest_score_From method, we wanted to test what would happen if the method was passed an empty array.  This is not expected input, and therefore, it is an edge case.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used max to get the highest scores, it was very useful because we did not have to write our own comparison method to find which score was higher by iterating through the hash.
What was one method you and your pair used to debug code? | We used awesome_print and pry to test the output of loops when we were getting results that were not what we expected.
Is there a specific piece of code you'd like feedback on? | The play word method in the Player class.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We learned a lot from working together, from methods and process to thinking about problems overall. We shared our opinions and listened to each other and tried really hard to work things out when we disagreed.  We did a good job of breaking to spike so that we could have a reference for further conversation when deciding how to move forward.

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,83 @@
...
+
+      return nil if letters.length > 7 || letters.length == 0
+
+      score += 50 if letters.length == 7
+
+      letters.each do |letter|
+        if !(letter =~ /[A-Z]/)
+          return nil
+        end
+        score += letter_score(letter)
```
Comment (CheezItMan) [inline lib/scoring.rb:36]: Good use of a helper method.

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,83 @@
...
+
+      max_keys = []
+      word_score.each do |key, value|
+        if value == word_score.values.max
+          max_keys << key
+        end
+      end
+
+      if max_keys.length == 1
+        return word_score.key(word_score.values.max)
```
Comment (CheezItMan) [inline lib/scoring.rb:62]: This is a little awkward.  since there is only one max word score, you could return,` max_keys.first`

File: lib/tile_bag.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        I: 9,	V: 2,
+        J: 1,	W: 2,
+        K: 1,	X: 1,
+        L: 4,	Y: 2,
+        M: 2,	Z: 1
+      }
+
+      letter_quantity.each do |letter, quantity|
+        quantity.times do
+          tiles << Scrabble::Tile.new(letter)
```
Comment (CheezItMan) [inline lib/tile_bag.rb:29]: Good choice putting the tiles into an array like this

File: lib/tile_bag.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        tiles_picked << some_tile.letter
+        tiles_remaining.delete(some_tile)
+      end
+
+      return tiles_picked
+    end
+
+  end
+
+  class Tile
```
Comment (CheezItMan) [inline lib/tile_bag.rb:54]: Think about POODR, don't write an additional class unless you need to.  This class really isn't doing anything more than a 1-letter string.

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          if !tiles.include?(char.to_sym)
+            return false
+          else
+            word_check << char.to_sym
+          end
+        end
+
+        plays << word
+        score = Scrabble::Scoring.score(word)
+        self.total_score += score
```
Comment (CheezItMan) [inline lib/player.rb:30]: This totally works, but you might consider making `total_score` a calculated field (a method) instead of an instance variable.

File: specs/tile_bag_spec.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  describe 'TileBag#draw_tiles(num)' do
+    it 'returns a collection of random tiles' do
+      # Arrange
+      test_1 = Scrabble::TileBag.new
+      # Act
+      result = test_1.draw_tiles(7)
+
+      # Assert
+      result.length.must_equal 7
+      result.wont_equal [:A, :A, :A, :A, :A, :A, :A]
```
Comment (CheezItMan) [inline specs/tile_bag_spec.rb:31]: This test is **unlikely**, but it's not impossible.

File: specs/tile_bag_spec.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      test_1.draw_tiles(5)
+
+      # Act
+      result = test_1.tiles_remaining.size
+
+      # Assert
+      result.must_equal 93
+    end
+
+    it 'returns tiles remaining if less than number of tiles requested' do
```
Comment (CheezItMan) [inline specs/tile_bag_spec.rb:46]: Good that you're testing this, I would also assert that `result` would be equal to 98.

File: specs/player_spec.rb [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+      the_test.play("camel")
+
+      the_test.tiles = [:F, :L, :I, :T, :T, :E, :R]
+      the_test.play("flitter")
+      # Act
+      result = the_test.highest_word_score
+
+      # Assert
+      result.must_equal 60
+    end
```
Comment (CheezItMan) [inline specs/player_spec.rb:184]: You should also have a `Player` test for drawing tiles.

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions |  Check, I'm glad you two worked together well
Both Teammates contributed to the codebase |  Both teammates have Commits
Regular Commits with meaningful commit messages | Good number of commits, but make the commit messages about the functionality added.
**Scoring class** |
`score` method | Check
Uses appropriate data structure to store the letter score |Check, although it could be more compact.
Appropriately handles edge cases | Check
`highest_score_from` method | Check, well done!
Appropriately handles edge cases | Check
Test implementations for `highest_score_from`| Check, well done
**Player class** |
Uses an array to keep track of words played | Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` | Check
Returns `true` or `false` from `won?` method | Check
Tests edge cases on the `play(word)` | Check
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag |  Well done
`draw_tiles` method uses the parameter to draw a set of random tiles |  Check
`tiles_remaining` method returns a count |Check
Tests `draw_tiles` edge cases | Check, good work testing the edge cases here.  You are missing tests for the player `draw_tiles` method
Summary | Very nicely done, you hit all the learning objectives.  Good work on testing edge-cases and good work with the `draw_tiles` and `highest_score_from` method.
